contracts/multisig-prover/src/execute.rs

```diff
@@ -126,31 +128,123 @@ pub fn update_worker_set(deps: DepsMut, env: Env) -> Result<Response, ContractEr
         pub_keys.push(pub_key);
     }
 
-    let new_worker_set = WorkerSet::new(snapshot, pub_keys, env.block.height)?;
+    Ok(WorkersInfo {
+        snapshot,
+        pubkeys_by_participant: participants.into_iter().zip(pub_keys).collect(),
+    })
+}
+
+pub fn update_worker_set(deps: DepsMut, env: Env) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+
+    let workers_info = get_workers_info(&deps, &env, &config)?;
+
+    let new_worker_set = WorkerSet::new(
+        workers_info.pubkeys_by_participant,
+        workers_info.snapshot.quorum.into(),
+        env.block.height,
+    )?;
 
-    let cur_worker_set = CURRENT_WORKER_SET.load(deps.storage)?;
+    let cur_worker_set = CURRENT_WORKER_SET.may_load(deps.storage)?;
 
-    if !should_update_worker_set(
-        &new_worker_set,
-        &cur_worker_set,
-        config.worker_set_diff_threshold as usize,
-    ) {
-        return Err(ContractError::WorkerSetUnchanged);
+    let key_id = new_worker_set.id();
+
+    match cur_worker_set {
+        None => {
+            // if no worker set, just store it and return
+            CURRENT_WORKER_SET.save(deps.storage, &new_worker_set)?;
+
+            KEY_ID.save(deps.storage, &key_id)?;
+
+            let key_gen_msg = multisig::msg::ExecuteMsg::KeyGen {
+                key_id,
+                snapshot: workers_info.snapshot,
+                pub_keys_by_address: new_worker_set
+                    .signers
+                    .into_iter()
+                    .map(|signer| {
+                        (
+                            signer.address.to_string(),
+                            (KeyType::Ecdsa, signer.pub_key.as_ref().into()),
+                        )
+                    })
+                    .collect(),
+            };
+
+            Ok(Response::new().add_message(wasm_execute(config.multisig, &key_gen_msg, vec![])?))
+        }
+        Some(cur_worker_set) => {
+            if !should_update_worker_set(
+                &new_worker_set,
+                &cur_worker_set,
+                config.worker_set_diff_threshold as usize,
+            ) {
+                return Err(ContractError::WorkerSetUnchanged);
+            }
+
+            NEXT_WORKER_SET.save(
+                deps.storage,
+                &(new_worker_set.clone(), workers_info.snapshot),
+            )?;
+            let mut builder = CommandBatchBuilder::new(config.destination_chain_id);
+            builder.add_new_worker_set(new_worker_set)?;
+
+            let batch = builder.build()?;
+
+            COMMANDS_BATCH.save(deps.storage, &batch.id, &batch)?;
+            REPLY_BATCH.save(deps.storage, &batch.id)?;
+
+            let start_sig_msg = multisig::msg::ExecuteMsg::StartSigningSession {
+                key_id: cur_worker_set.id(), // TODO remove the key_id
+                msg: batch.msg_to_sign(),
+            };
+
+            Ok(Response::new().add_submessage(SubMsg::reply_on_success(
```

Comment:
what about failure case?

Reply:
do we want to handle the failure? I think it makes more sense if it reverts which is the case for failed submessages set to reply on success

Reply:
good to know 👍